lib/nat/service.go

```diff
@@ -206,41 +210,36 @@ func (s *Service) RemoveMapping(mapping *Mapping) {
 // Optionally takes renew flag which indicates whether or not we should renew
 // mappings with existing natds
 func (s *Service) updateMapping(ctx context.Context, mapping *Mapping, nats map[string]Device, renew bool) {
-	var added, removed []Address
-
 	renewalTime := time.Duration(s.cfg.Options().NATRenewalM) * time.Minute
-	mapping.expires = time.Now().Add(renewalTime)
 
-	newAdded, newRemoved := s.verifyExistingMappings(ctx, mapping, nats, renew)
-	added = append(added, newAdded...)
-	removed = append(removed, newRemoved...)
+	mapping.mut.Lock()
 
-	newAdded, newRemoved = s.acquireNewMappings(ctx, mapping, nats)
-	added = append(added, newAdded...)
-	removed = append(removed, newRemoved...)
+	mapping.expires = time.Now().Add(renewalTime)
+	change := s.verifyExistingMappings(ctx, mapping, nats, renew)
+	add := s.acquireNewMappings(ctx, mapping, nats)
 
-	if len(added) > 0 || len(removed) > 0 {
-		mapping.notify(added, removed)
+	mapping.mut.Unlock()
+
+	if change || add {
+		mapping.notify()
 	}
 }
 
-func (s *Service) verifyExistingMappings(ctx context.Context, mapping *Mapping, nats map[string]Device, renew bool) ([]Address, []Address) {
-	var added, removed []Address
-
+func (s *Service) verifyExistingMappings(ctx context.Context, mapping *Mapping, nats map[string]Device, renew bool) (change bool) {
```

Comment:
Shouldn't these also be "locked"?

Reply:
Done. I did s/Mappings/Locked/, because in the terminology of this package, each handles a single Mapping.